core/src/SYCL/Kokkos_SYCL_Instance.hpp

```diff
@@ -70,18 +70,206 @@ class SYCLInternal {
 
   std::unique_ptr<sycl::queue> m_queue;
 
+  // USMObjectMem is a reusable buffer for a single object
+  // in USM memory
+  template <sycl::usm::alloc Kind>
+  class USMObjectMem {
+   public:
+    class Deleter {
+     public:
+      Deleter() = default;
+      explicit Deleter(USMObjectMem* mem) : m_mem(mem) {}
+
+      template <typename T>
+      void operator()(T* p) const noexcept {
+        assert(m_mem);
+        assert(sizeof(T) == m_mem->m_size);
+
+        if constexpr (sycl::usm::alloc::device == kind)
+          // Only skipping the detor on trivially copyable types
+          static_assert(std::is_trivially_copyable_v<T>);
+        else
+          p->~T();
+
+        m_mem->m_size = 0;
+      }
+
+     private:
+      USMObjectMem* m_mem = nullptr;
+    };
+
+    static constexpr sycl::usm::alloc kind = Kind;
+
+    void reset() {
+      assert(!m_size);
+
+      if (m_data) {
+        sycl::free(m_data, *m_q);
+        m_capacity = 0;
+        m_data     = nullptr;
+      }
+      m_q.reset();
+    }
+
+    void reset(sycl::queue q) {
+      reset();
+      m_q.emplace(std::move(q));
+    }
+
+    USMObjectMem() = default;
+    explicit USMObjectMem(sycl::queue q) noexcept : m_q(std::move(q)) {}
+
+    USMObjectMem(USMObjectMem const&) = delete;
+    USMObjectMem(USMObjectMem&&)      = delete;
+    USMObjectMem& operator=(USMObjectMem&&) = delete;
+    USMObjectMem& operator=(USMObjectMem const&) = delete;
+
+    ~USMObjectMem() {
+      assert(!m_size);
+
+      if (m_data) sycl::free(m_data, *m_q);
```

Comment:
Might want to `assert(m_q)`

Reply:
These asserts are really meant as precondition checks on external calls.  In this particular case, what is being checked by the assert is that there not an outstanding `unique_ptr` (`m_data` doesn't point to a live object).  There is no (external) way for `m_q` to not be engaged when `m_data` is set (a bug completely internal to `USMObjectMem` could, of course, cause this, as that would mean the invariants are broken), which is why I don't check it.